### PR TITLE
Debug New Motion 

### PR DESCRIPTION
**Pull request by pvillacorta**
This PR is intended to debug the error of ArbitraryAction motion when testing in oneAPI. 
I don't want to fill #442 with (more) unnecessary commits, so I'll upload them here and delete the PR when I find the error.

### CODE REVIEW

File: KomaMRIBase/src/motion/motionlist/actions/ArbitraryAction.jl [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    action::ArbitraryAction{T},
+    x::AbstractArray{T},
+    y::AbstractArray{T},
+    z::AbstractArray{T},
+    t::AbstractArray{T},
+) where {T<:Real}
+    itp = interpolate(action.dy, Gridded(Linear()), Val(size(action.dy,1)))
+    uy .= resample(itp, t)
+    return nothing
+end
```
Comment (cncastillo) [inline KomaMRIBase/src/motion/motionlist/actions/ArbitraryAction.jl:83]: Try this:

```suggestion
function displacement_y!(
    uy::AbstractArray{T},
    action::ArbitraryAction{T},
    x::AbstractArray{T},
    y::AbstractArray{T},
    z::AbstractArray{T},
    t::AbstractArray{T},
) where {T<:Real}
    itp = interpolate(action.dy, Gridded(Linear()), Val(size(action.dy,1)))
    uy_s = resample(itp, t)
    uy .= uy_s
    return nothing
end
```

File: KomaMRICore/src/simulation/Functors.jl [hunk 2]
```diff
@@ -52,9 +51,9 @@ See also [`f32`](@ref) and [`f64`](@ref) to change element type only.
 x = gpu(x, CUDABackend())
 ```
 """
-function gpu(x, backend::KA.GPU)
-    return fmap(x -> adapt(backend, x), x; exclude=_isleaf)
-end
+gpu(x, backend::KA.GPU) = fmap(x -> adapt(backend, x), x; exclude=_isleaf)
+adapt_storage(backend::KA.GPU, xs::MotionList) = MotionList(gpu.(xs.motions, Ref(backend)))
+adapt_storage(backend::KA.GPU, xs::Motion) = Motion(gpu(xs.action, backend), gpu(xs.time, backend), xs.spins)
```
Comment (cncastillo) [inline KomaMRICore/src/simulation/Functors.jl:56]: This lines should be here. All the Koma-related parts are at the end.

File: KomaMRICore/ext/KomaoneAPIExt.jl [hunk 3]
```diff
@@ -62,8 +62,8 @@ function __init__()
 end
 
 ## Extend KomaMRIBase.unit_time (until bug with oneAPI is solved)
-KomaMRIBase.unit_time(t::OneVector, ts::TimeRange) = begin
```
Comment (cncastillo) [inline KomaMRICore/ext/KomaoneAPIExt.jl]: This was not the solution, at the top of the file, use `using KomaMRICore`. `KomaMRIBase.unit_time` was correct.

File: KomaMRICore/src/simulation/Functors.jl (hunk 2 above)
Comment (pvillacorta) [reply KomaMRICore/src/simulation/Functors.jl]: Solved in [ee18d14](https://github.com/JuliaHealth/KomaMRI.jl/pull/472/commits/ee18d14dc9a047f7a87058bddf37ac8e537e8edb)

**Comment by pvillacorta** [comment]
I am not able to reproduce the error, as it does not appear when I try to isolate it. I will try again, but, for the moment, I merge this branch into #442 in order to have, at least, a temporary solution for this.

**Comment by maleadt** [comment]
The CI jobs here have killed our oneAPI CI server _multiple times_, consuming over 20GB of shared memory, so I'm temporarily disabling the JuliaGPU Buildkite integration KomaMRI.jl.

**Comment by cncastillo** [comment]
I think I am able to re-enable it myself. Thank you!

Talking with @pvillacorta, it seems that debugging this locally would be hard, as it appears that we encountered an issue with oneAPI.jl. This seems to be a problem with `max/min` like the one we had on Metal.jl  https://github.com/JuliaGPU/GPUCompiler.jl/pull/602/files#diff-52d57b551d62b4dc18df904b53dfbbf149404dd5f9e748b0f417364f5d32966f. On oneAPI.jl the symptoms are different, instead of being wrong all the time like for Metal.jl, the operation on oneAPI.jl **sometimes gives incorrect results** (approx 2/10 times). 

```julia
function _unit_time(t::AbstractArray{T}, ts::TimeRange{T}) where {T<:Real}
    if ts.t_start == ts.t_end
        return (t .>= ts.t_start) .* oneunit(T)
    else
        # Metal.jl: Separating the min/max statements fixes the problem for now
        # until https://github.com/JuliaGPU/GPUCompiler.jl/pull/602/files#diff-52d57b551d62b4dc18df904b53dfbbf149404dd5f9e748b0f417364f5d32966f
        tmp = max.((t .- ts.t_start) ./ (ts.t_end - ts.t_start), zero(T))
        # oneAPI.jl: The problem was fixed by adding an additional instruction here
        # _ = sum(t)
        # Could this be a GPUCompiler.jl problem?
        return min.(tmp, oneunit(T))
    end
end
```
Pablo hasn't completely identified/isolated the problem, as he tells me that it is a little bit random when it happens, so he hasn't filed an issue.

@maleadt Do you think using `ifelse` instead of `min/max` could be better? Would you happen to have any recommendations for what to do?

Thank you very much for all your help!

**Comment by pvillacorta** [comment]
@maleadt, I want to apologize for the inconvenience I've caused with my recent commits. It was never my intention to overload the server or create issues in the CI environment. I’m working on debugging this issue as best as I can, but I’ve encountered some challenges, as @cncastillo mentioned.

I understand the importance of being more mindful of shared resources and I’m committed to changing my approach to ensure that I don’t cause further disruptions. I would greatly appreciate any advice or guidance on how I could handle this situation more effectively. Once again, I’m very sorry for what happened and I’m eager to collaborate on resolving this in the best possible way.

Thank you for your understanding and help.

**Comment by maleadt** [comment]
> debugging this locally would be hard, as it appears that we encountered an issue with oneAPI.jl.

Is the issue that you don't have the hardware? Many Intel CPUs come with an oneAPI-compatible IGP.


> On oneAPI.jl the symptoms are different, instead of being wrong all the time like for Metal.jl, the operation on oneAPI.jl **sometimes gives incorrect results** (approx 2/10 times).

The function you quote there doesn't look like a kernel issue, so it's probably different. If this ends up dispatching to MKL, this may be https://github.com/JuliaGPU/oneAPI.jl/issues/445; try setting the environment variable mentioned in that issue.

> Do you think using `ifelse` instead of `min/max` could be better?

If this is a kernel function (which it doesn't look like), using `ifelse` reduces branching which may help.

**Comment by pvillacorta** [comment]
From [c71fee8](https://github.com/JuliaHealth/KomaMRI.jl/pull/472/commits/c71fee88b202b47aa4ab4da39d32c7177919aba0), we can notice how [tests fail](https://buildkite.com/julialang/komamri-dot-jl/builds/1090#0191b70d-9649-423a-991c-60c09884425d) even using `KernelAbstractions.synchronize` after calculating the time array:
```julia
const AdjointOneArray{T, N, M} = LinearAlgebra.Adjoint{T, oneArray{T, N, M}} where {T, N, M}
## Extend KomaMRIBase.unit_time (until bug with oneAPI is solved)
function KomaMRIBase.unit_time(t::AdjointOneArray{T, N, M}, ts::KomaMRIBase.TimeRange{T}) where {T, N, M}
    if ts.t_start == ts.t_end
        return (t .>= ts.t_start) .* oneunit(T)
    else
        tmp = max.((t .- ts.t_start) ./ (ts.t_end - ts.t_start), zero(T))
        t = min.(tmp, oneunit(T))
        KA.synchronize(KA.get_backend(t))
        return t
    end
end
```

What I thought would solve the problem every time, was to use a somewhat heavier operation that would make use of the calculated variable t, such as a `sum` function (see https://github.com/JuliaHealth/KomaMRI.jl/commit/97048f063f3d89ad2969e497511349123199d83f). [The tests](https://buildkite.com/julialang/komamri-dot-jl/builds/1115#0191b83e-b669-464e-b8b9-7717346dff43) now pass 8/8 times, but it may be a coincidence:
```julia
## Extend KomaMRIBase.unit_time (until bug with oneAPI is solved)
function KomaMRIBase.unit_time(t::AdjointOneArray{T, N, M}, ts::KomaMRIBase.TimeRange{T}) where {T<:Real, N, M}
    if ts.t_start == ts.t_end
        return (t .>= ts.t_start) .* oneunit(T)
    else
        tmp = max.((t .- ts.t_start) ./ (ts.t_end - ts.t_start), zero(T))
        t = min.(tmp, oneunit(T))
        _ = sum(t) # Dummy sum
    end
end
```

As I thought that this sum could be a good temporary solution (until I could find the source of the problem), I included it in branch #442 (commit [51c71cc](https://github.com/JuliaHealth/KomaMRI.jl/pull/442/commits/51c71cca2d12297ede3c722c8d03cb731d0460cc)), but [the test](https://buildkite.com/julialang/komamri-dot-jl/builds/1114#_) failed again for oneAPI.

So I started to think that this function `unit_time` was not the source of the problem, and I tried to isolate and reproduce the problem (commits [3e5b845](https://github.com/JuliaHealth/KomaMRI.jl/pull/472/commits/3e5b84545010cb22da496bba70fcd588e4756e06) in advance) including also Interpolations.jl, to see if it was something related with this package. I did not get it, by the way, as all tests were passing now.

**Comment by pvillacorta** [comment]
> > debugging this locally would be hard, as it appears that we encountered an issue with oneAPI.jl.
> 
> Is the issue that you don't have the hardware? Many Intel CPUs come with an oneAPI-compatible IGP.
> 
> > On oneAPI.jl the symptoms are different, instead of being wrong all the time like for Metal.jl, the operation on oneAPI.jl **sometimes gives incorrect results** (approx 2/10 times).
> 
> The function you quote there doesn't look like a kernel issue, so it's probably different. If this ends up dispatching to MKL, this may be [JuliaGPU/oneAPI.jl#445](https://github.com/JuliaGPU/oneAPI.jl/issues/445); try setting the environment variable mentioned in that issue.
> 
> > Do you think using `ifelse` instead of `min/max` could be better?
> 
> If this is a kernel function (which it doesn't look like), using `ifelse` reduces branching which may help.

Just saw this message, I will check it and let you know.
Thank you!

**Pull request closed without merging**